### DISCUSSION TITLE
How to decode with CUDA hardware instead of CPU?

### DISCUSSION THREAD

**Discussion by Cadene** [python-ffmpegio/python-ffmpegio#43] [Q&A]
This is what I use to decode with CPU. How do I decode with CUDA hardware?
```python
fs, frames = ffmpegio.video.read(str(video_path), ss=str(timestamp), pix_fmt=image_format)
```

Thanks 🙏

**Comment by tikuma-lsuhsc** [comment]
The short answer is "as you do so in FFmpeg" but it may not be helpful if you're not familiar with it...

Per [FFmpeg Wiki](https://trac.ffmpeg.org/wiki/HWAccelIntro#NVDECCUVID), you need to invoke FFmpeg with `-hwaccel cuda` option like:

```
ffmpeg -hwaccel cuda -i input output
```

When you call any `ffmpegio` reader/writer functions, you can just add any FFmpeg ouput/global option as a keyword argument. So, in your case all you need to do is:

<strike>
fs, frames = ffmpegio.video.read(str(video_path), ss=str(timestamp), pix_fmt=image_format, hwaccel='cuda')
</strike>

```python
fs, frames = ffmpegio.video.read(str(video_path), ss=str(timestamp), pix_fmt=image_format, hwaccel_in='cuda')
```

(edit: `hwaccel` is an FFmpeg input option. Thus, you must use `hwaccel_in` keyword input in `ffmpegio`)


(To add input option, you need to add `_in` to the option name FYI)

**Comment by gyeongjin-kim** [comment]
How to apply this way in `ffmpegio.open()` ?
I have tried to add `hwaccel='cuda'` into a `open()` as a argument, but it couldn't work to sample block..

Following is the code.
```
with ffmpegio.open(
                self.hls_url, "rva", blocksize=BLOCK_SIZE,
                s=(640, 360), ac=1, sample_rate=SAMPLE_RATE, sample_fmt="flt", hwaccel='cuda'
            ) as stream:
```

**Reply by tikuma-lsuhsc** [reply]
I'll look into it closer later, but I think I made a mistake above. It should be `hwaccel_in` as `hwaccel` FFmpeg option is an input option. Also, could you fill in on what you mean by "it couldn't work to sample block"? If there is an error message, please post it. Thanks.

**Reply by tikuma-lsuhsc** [reply]
Confirmed.

```python
import ffmpegio as ff

hls_url = "tests/assets/sample.mp4"
BLOCK_SIZE = 10
SAMPLE_RATE = 8000

with ff.open(
    hls_url,
    "rva",
    blocksize=BLOCK_SIZE,
    s=(640, 360),
    ac=1,
    sample_rate=SAMPLE_RATE,
    sample_fmt="flt",
    hwaccel_in="cuda",
    show_log=True,
    loglevel='debug'
) as stream:
    for f in stream: # audio & video stream raw data are stored in a dict with stream ids as the keys
        print(f.keys()) 
        exit()
```

If you turn on `show_log` and set the FFmpeg `loglevel` to `'debug'`, it confirms that you are loading the NV decoder:

```
[AVHWDeviceContext @ 000002a7c70f7e40] Loaded lib: nvcuda.dll
[AVHWDeviceContext @ 000002a7c70f7e40] Loaded sym: cuInit
[AVHWDeviceContext @ 000002a7c70f7e40] Loaded sym: cuDriverGetVersion
[AVHWDeviceContext @ 000002a7c70f7e40] Loaded sym: cuDeviceGetCount
[AVHWDeviceContext @ 000002a7c70f7e40] Loaded sym: cuDeviceGet
[AVHWDeviceContext @ 000002a7c70f7e40] Loaded sym: cuDeviceGetAttribute
...
```

`hwaccel_in` must be used instead of `hwaccel` or else read errors out due to:

```
Option hwaccel (use HW accelerated decoding) cannot be applied to output url - -- you are trying to apply an input option to an output file or vice versa. Move this option before the file it belongs to.
```

**Reply by gyeongjin-kim** [reply]
Thank you. It's working properly. I couldn't see that in the official documentation, so it would be nice to see it added.

**Reply by tikuma-lsuhsc** [reply]
👍  

[It is actually mentioned in the documentation](https://python-ffmpegio.github.io/python-ffmpegio/options.html#ffmpeg-option-references)... but yeah, I'm expecting the users to be well-versed in FFmpeg options, but they are definitely not most obvious for those who have not used it before. Maybe I should create a FAQ covering hardware acceleration...
